core/src/main/java/org/testcontainers/containers/wait/strategy/HostPortWaitStrategy.java

```diff
@@ -41,10 +49,52 @@ protected void waitUntilReady() {
         Callable<Boolean> externalCheck = new ExternalPortListeningCheck(waitStrategyTarget, externalLivenessCheckPorts);
 
         try {
-            Unreliables.retryUntilTrue((int) startupTimeout.getSeconds(), TimeUnit.SECONDS,
-                () -> getRateLimiter().getWhenReady(() -> internalCheck.call() && externalCheck.call()));
+            List<Future<Boolean>> futures = EXECUTOR.invokeAll(Arrays.asList(
+                // Blocking
+                () -> {
+                    Instant now = Instant.now();
+                    Boolean result = internalCheck.call();
+                    log.debug(
+                        "Internal port check {} for {} in {}",
+                        Boolean.TRUE.equals(result) ? "passed" : "failed",
+                        internalPorts,
+                        Duration.between(now, Instant.now())
+                    );
+                    return result;
+                },
+                // Polling
+                () -> {
+                    Instant now = Instant.now();
+                    RateLimiter rateLimiter = RateLimiterBuilder
```

Comment:
WDYT about replacing duct-tape with awaitility here?

Reply:
Apparently, Awaitility was performing much slower (~100ms vs 10ms). Need to investigate why (and maybe revisit our previous plans to migrate)

Reply:
Not sure if directly related to your observation, but the way Awaitility handles its executors by default might introduce some overhead, see https://github.com/awaitility/awaitility/issues/168.

Reply:
I am an idiot - I missed the fact that, unless explicitly set, `pollDelay` is equal to `pollInterval`, so it was adding a static delay 😅 With `pollDelay` set to `0`, Awaitility is as fast as `while (...) {}` 🎉

Reply:
Ouh yeah! But this is easy to get wrong, not the most intuitive default IMO.